design/weak-imports.md

```diff
@@ -0,0 +1,60 @@
+Weak Imports
+============
+
+It can be useful for WASI programs to weakly depend on certain functions in a
+WASI API.  For example, if a WASI API evolves to include a new function a
+program might want to continue to run on older systems that don't yet support
+the new function.  In this case a weak import mechanism allows the program to
+run on older systems and detect the presence of the function at runtime.
+
+WebAssembly itself does not currently provide a mechanism for weak imports.
+There is some discussion of adding it the spec [^1], and WASI would
+likely use such a feature if/when it is added.
+
+This document describes the mechanism used by WASI to achieve a form of weak
+import on top of the existing primitives.  Currently this is only defined for
+function imports.
+
+Declaring a weak function import
+--------------------------------
+
+Weak function imports are implemened by using two imports for each function.
+The first being the weak function and the second being a WebAssembly i32 global
+to indicate if the function is present at runtime.
+
+For example, if a module called `wasi:fs` added a new `statvfs` function to its
+interface a program could import this new function weakly in the following way:
+
+```wasm
+(func $statvfs (import "wasi:fs" "statvfs.weak"))
+(global $statvfs_is_present (import "wasm:fs" "statvfs.is_present") i32)
```

Comment:
What if we put the ".weak" and ".is_present" in the module name, rather than the field name? This is a little less intuitive, however it would address the concern about intruding on what would otherwise be the user's or compiler's namespace. `(import "wasi:fs.weak" "statvfs")` and `(import "wasi:fs.is_present" "statvfs")`?

Reply:
Interesting.  Kind of nasty but see you point.   What do others think of this idea?

Reply:
Another option would be to define a reserved character to use instead of `.`. It's unlikely any system would want to use `\0` for its symbol names, for example.

Reply:
Is there some existing encoding that we could reuse here, e.g. url encoding?

Reply:
Using null bytes in strings seems a like asking for trouble.

Reply:
If any character (combination), then definitely something visible in ASCII.